src/Stateless/Reflection/MethodInfo.cs

```diff
@@ -0,0 +1,54 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Stateless.Reflection
+{
+    /// <summary>
+    /// Describes an internal method through the reflection API.
+    /// </summary>
+    public class MethodInfo
+    {
+        static readonly string _defaultFunctionName = "Function";
+
+        internal static MethodInfo CreateMethodInfo(MethodDescription action)
+        {
+            string descrip;
+
+            descrip = action.Description;
+
+            // If the action description was calculated based on the method name, and the method was a delegate or lambda,
+            // don't use the compiler-generated method name.
+            if ( (action.DescriptionIsCalculated) && (descrip.IndexOfAny(new char[] { '<', '>' }) >= 0) )
+                descrip = _defaultFunctionName;
+
+            MethodInfo actionInfo = new MethodInfo(descrip, action.IsAsync);
```

Comment:
I simplified this even further, there's no reason to have that intermediate value at all, just return the new'ed MethodInfo.